File: scripts/ci/sigstack_fix.c
/*
 * uplink.NET sigaltstack fix — prototype.
 *
 * Hypothesis we are testing (see docs/crash-investigation.md):
 *   When a .NET thread has done a P/Invoke into libstorj_uplink.so,
 *   Go's cgo runtime installs a ~32 KB sigaltstack on that thread.
 *   After the P/Invoke returns, the alt stack stays set. Later, when
 *   CoreCLR's JIT emits a managed null-reference probe (deliberate
 *   deref of address 0, handled via SIGSEGV), the kernel delivers
 *   SIGSEGV on Go's small alt stack. CoreCLR's SIGSEGV handler is
 *   complex and overflows that alt stack, hitting the guard page and
 *   producing si_code=SI_KERNEL — the exact signature seen in three
 *   captured cores.
 *
 * This shim exposes a single entry point that installs a large (1 MB)
 * sigaltstack on the calling thread. The managed side calls it once
 * per thread (gated by ThreadStatic) from Access.AcquireProjectLease
 * and related entry points, before any P/Invoke into libstorj_uplink.
 * Once set, the 1 MB alt stack persists for the life of the thread,
 * leaving enough headroom for CoreCLR's handler to run without
 * overflow.
 *
 * The buffer is leaked intentionally: it must outlive any signal that
 * uses it, which on .NET's thread pool means "for the life of the
 * thread". 1 MB × ~200 pool threads max is well under 256 MB and
 * strictly bounded.
 *
 * Build (Linux):
 *   cc -O2 -fPIC -shared -o libuplink_sigstack_fix.so sigstack_fix.c
 */

#include <signal.h>
#include <stdlib.h>
#include <string.h>

/*
 * Must comfortably exceed what CoreCLR's SIGSEGV handler walks in the
 * worst case: managed stack unwind, type handle lookups, exception
 * object construction, a call out to createdump if DOTNET_DbgEnableMiniDump
 * is set. Empirically ~128 KB is enough; we use 1 MB for a generous
 * safety margin.
 */
#define UPLINK_SIGALTSTACK_SIZE (1024 * 1024)

/*
 * Install a large sigaltstack on the calling thread. Returns 0 on
 * success, a negative errno on failure. Safe to call more than once
 * on the same thread; only the first call has any effect.
 *
 * The function is idempotent-per-thread via an explicit thread-local
 * flag so the managed side can call it unconditionally at the top of
 * hot paths without double-allocating.
 */
/*
 * Thread-local pointer to the large alt stack we allocated. NULL if
 * this thread has never had one installed. Re-used on subsequent
 * calls if Go's cgo runtime overwrote it with its smaller stack.
 */
static __thread void *our_stack = NULL;

__attribute__((visibility("default")))
int uplink_sigstack_install(void) {
    /*
     * Check the ACTUAL current sigaltstack rather than relying on a
     * boolean "installed" flag. Previous iteration used a fire-once
     * flag and the fix was silently overridden: we installed before
     * the cgo call, Go's minitSignalStack() replaced it with a ~32 KB
     * stack during the first cgo entry on that thread, and the flag
     * prevented us from re-installing afterwards.
     *
     * sigaltstack(NULL, &current) is cheap — reads from the kernel
     * task struct, no context switch on Linux ≥ 5.x.
     */
    stack_t current;
    if (sigaltstack(NULL, &current) == 0
            && !(current.ss_flags & SS_DISABLE)
            && current.ss_size >= UPLINK_SIGALTSTACK_SIZE) {
        return 0; /* already large enough, whether ours or someone else's */
    }

    /* Allocate once per thread; re-use on subsequent calls. */
    if (our_stack == NULL) {
        our_stack = malloc(UPLINK_SIGALTSTACK_SIZE);
        if (our_stack == NULL) {
            return -1;
        }
        /* Touch every page so the kernel commits them; avoids a soft
         * fault *during* signal delivery. */
        memset(our_stack, 0, UPLINK_SIGALTSTACK_SIZE);
    }

    stack_t ss;
    ss.ss_sp    = our_stack;
    ss.ss_size  = UPLINK_SIGALTSTACK_SIZE;
    ss.ss_flags = 0;

    if (sigaltstack(&ss, NULL) != 0) {
        return -2;
    }

    /* Intentionally leak `our_stack` — the kernel holds a reference
     * for signal delivery that persists for the life of this thread. */
    return 0;
}

File: src/uplink.NET/Native/SigStackFix.cs
using System.Runtime.InteropServices;

namespace uplink.NET.Native;

/// <summary>
/// Prototype workaround for the intermittent Linux test-host SIGSEGV
/// documented in <c>docs/crash-investigation.md</c>.
///
/// Installs a 1 MB <c>sigaltstack</c> on every thread that is about to
/// call into <c>libstorj_uplink.so</c>. Go's cgo runtime sets a ~32 KB
/// alt stack on cgo-entered threads and that alt stack persists after
/// the cgo call returns. When CoreCLR later delivers SIGSEGV on that
/// thread (via its managed null-reference probe machinery), CoreCLR's
/// handler overflows the small alt stack and produces a kernel-synthesized
/// SIGSEGV (<c>si_code=SI_KERNEL</c>) — the exact signature seen in
/// three captured cores.
///
/// Ensuring every cgo-touched thread has a 1 MB alt stack gives CoreCLR's
/// handler enough headroom to run without overflow. We pay for this
/// once per thread (<see cref="_installed"/> is <see cref="ThreadStaticAttribute"/>)
/// and the backing buffer is intentionally leaked for the life of the
/// thread because the kernel holds a reference to it for signal delivery.
///
/// This file is prototype scaffolding: if the CI evidence confirms it
/// removes the crash, we'll either promote it to a proper library
/// feature, push it upstream to <c>dotnet/runtime</c> / <c>golang/go</c>
/// as a known-bad-interaction bug, or both.
/// </summary>
internal static partial class SigStackFix
{
    private const string LibName = "uplink_sigstack_fix";

    // True once we've confirmed the shim is loadable. Once it fails
    // with DllNotFoundException, we stop trying on ALL threads.
    [ThreadStatic]
    private static bool _checkedOnThisThread;
    private static volatile bool _shimUnavailable;

    /// <summary>
    /// Ensure the current thread has a large sigaltstack. Designed to
    /// be called on every cgo entry point (before AND after the
    /// P/Invoke). Cheap: the native side reads the actual sigaltstack
    /// size from the kernel task struct and returns immediately if it's
    /// already large enough.
    ///
    /// Must be called AFTER the first cgo call on a thread too, not
    /// just before, because Go's cgo runtime replaces whatever
    /// sigaltstack was set with its own ~32 KB stack during M init.
    /// </summary>
    internal static void EnsureOnCurrentThread()
    {
        if (_shimUnavailable)
            return;

        try
        {
            uplink_sigstack_install();
            _checkedOnThisThread = true;
        }
        catch (DllNotFoundException)
        {
            _shimUnavailable = true;
        }
    }

    [LibraryImport(LibName)]
    private static partial int uplink_sigstack_install();
}
